File: codes/matlab/ch20/laguerre_vs_tln.m
function laguerre_vs_tln()
%% laguerre_vs_tln - Chapter 20, Etude 20.5.  Laguerre vs TL_n.
% Author: Dr. Denys Dutykh.

    script_dir = fileparts(mfilename('fullpath'));
    addpath(fullfile(script_dir, '..', 'ch07'));
    out_dir = fullfile(script_dir, '..', '..', '..', ...
                       'textbook', 'figures', 'ch20', 'matlab');
    if ~exist(out_dir, 'dir'); mkdir(out_dir); end

    Ns = [8 12 16 24 32 48 64 96];
    err_lag_e = arrayfun(@(N) lag_err(@(y) exp(-y), N), Ns);
    err_lag_a = arrayfun(@(N) lag_err(@(y) 1./(1+y), N), Ns);
    err_tln_e = arrayfun(@(N) tln_err(@(y) exp(-y), N, 2.0), Ns);
    err_tln_a = arrayfun(@(N) tln_err(@(y) 1./(1+y), N, 5.0), Ns);

    NAVY=[20 45 110]/255; CORAL=[231 76 60]/255; TEAL=[22 160 133]/255;
    fig = figure('Position',[100 100 1020 340],'Color','w');

    subplot(1,2,1);
    semilogy(Ns, err_lag_e + 1e-18, '-o', 'Color', CORAL); hold on;
    semilogy(Ns, err_tln_e + 1e-18, '-s', 'Color', TEAL);
    xlabel('N'); ylabel('max error'); grid on; box on;
    title('(a) f(y) = e^{-y} (exponential)');
    legend({'Laguerre','TL_n L=2'});

    subplot(1,2,2);
    semilogy(Ns, err_lag_a + 1e-18, '-o', 'Color', CORAL); hold on;
    semilogy(Ns, err_tln_a + 1e-18, '-s', 'Color', TEAL);
    xlabel('N'); ylabel('max error'); grid on; box on;
    title('(b) f(y) = 1/(1+y) (algebraic)');
    legend({'Laguerre','TL_n L=5'});

    print(fig, fullfile(out_dir, 'laguerre_vs_tln.pdf'), '-dpdf');
    print(fig, fullfile(out_dir, 'laguerre_vs_tln.png'), '-dpng');
    fprintf('[20.5-matlab] figure saved\n');
end

function L = lag_poly(n, y)
    y = y(:)';
    L = zeros(n+1, length(y));
    L(1,:) = 1;
    if n >= 1, L(2,:) = 1 - y; end
    for k = 1:n-1
        L(k+2,:) = ((2*k+1 - y) .* L(k+1,:) - k * L(k,:)) / (k+1);
    end
end

function e = lag_err(f, N)
    [x, w] = gauss_laguerre(N + 32);
    L = lag_poly(N, x);
    fv = f(x);
    c = sum(w .* exp(x/2) .* fv .* L, 2);
    y = linspace(0.001, 60, 4001);
    Ly = lag_poly(N, y);
    approx = exp(-y/2) .* sum(c .* Ly, 1);
    e = max(abs(approx - f(y)));
end

function e = tln_err(f, N, L)
    [~, x] = cheb_matrix(N);
    y = L * (1 + x) ./ (1 - x);
    fv = zeros(size(y));
    ok = abs(x) < 1 - 1e-12;
    fv(ok) = f(y(ok));
    a = dct1(fv);
    y_fine = linspace(0.001, 60, 4001);
    xf = (y_fine - L) ./ (y_fine + L);
    approx = cheb_eval(a, xf, N);
    e = max(abs(approx - f(y_fine)));
end

function [x, w] = gauss_laguerre(K)
    n = K; i = 1:n-1;
    alpha_i = (2*(0:n-1) + 1);
    beta_i = i;
    J = diag(alpha_i) + diag(sqrt(beta_i.*beta_i), 1) + diag(sqrt(beta_i.*beta_i), -1);
    [V, D] = eig(J);
    x = diag(D)';
    w = V(1,:).^2;
end

function a = dct1(v)
    v = v(:);
    N = length(v) - 1;
    V = [v; v(N:-1:2)];
    A = real(fft(V))/N; A(1) = A(1)/2; A(N+1) = A(N+1)/2;
    a = A(1:N+1);
end

function v = cheb_eval(a, x, N)
    T0 = ones(size(x)); T1 = x;
    v = a(1)*T0 + a(2)*T1;
    for n = 2:N
        Tk = 2*x.*T1 - T0;
        v = v + a(n+1)*Tk;
        T0 = T1; T1 = Tk;
    end
end

File: codes/julia/ch20/laguerre_vs_tln.jl
# laguerre_vs_tln.jl
# Chapter 20: Spectral Methods on Unbounded Intervals
# Computational Etude 20.5: Laguerre vs TL_n on [0, infty).
#
# Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)

using CairoMakie
using FFTW
using LinearAlgebra
using Printf

include(joinpath(@__DIR__, "..", "ch07", "cheb_matrix.jl"))

function laguerre_poly(n, y)
    yv = collect(y)
    ell = zeros(n + 1, length(yv))
    ell[1, :] .= 1.0
    if n >= 1
        ell[2, :] .= 1.0 .- yv
    end
    for k in 1:(n - 1)
        ell[k + 2, :] .= ((2k + 1 .- yv) .* ell[k + 1, :] .- k .* ell[k, :]) ./ (k + 1)
    end
    return ell
end

# Golub-Welsch Gauss-Laguerre
function gauss_laguerre(K)
    a = collect(2 .* (0:K - 1) .+ 1.0)
    b = collect(1.0:K - 1)
    J = diagm(0 => a, 1 => b, -1 => b)
    vals, vecs = eigen(J)
    x = real(vals)
    w = real.(vecs[1, :] .^ 2)
    return x, w
end

function laguerre_error(f, N)
    x, w = gauss_laguerre(N + 32)
    ell = laguerre_poly(N, x)
    fv = f.(x)
    # a_n = int f(y) exp(-y/2) L_n(y) dy, with Gauss-Laguerre Gauss-Laguerre integrates e^{-x} g(x)
    # so we insert e^{x/2}: a_n = sum_i w_i exp(x_i) * f(x_i) e^{-x_i/2} L_n(x_i) = sum w_i e^{x_i/2} f(x_i) L_n(x_i)
    c = vec(sum(ell .* (w .* exp.(x ./ 2) .* fv)', dims=2))
    y = collect(range(0.001, 60.0, length=4001))
    Ly = laguerre_poly(N, y)
    approx = exp.(-y ./ 2) .* vec(sum(c .* Ly, dims=1))
    return maximum(abs.(approx .- f.(y)))
end

function dct1_coeffs(v)
    N = length(v) - 1
    V = vcat(v, reverse(v[2:N]))
    A = real.(fft(V)) ./ N
    A[1] *= 0.5; A[N + 1] *= 0.5
    return A[1:N + 1]
end

function cheb_eval(a, x, N)
    T0 = ones(length(x)); T1 = copy(x)
    val = a[1] .* T0 .+ a[2] .* T1
    for n in 2:N
        Tk = 2 .* x .* T1 .- T0
        val .+= a[n + 1] .* Tk
        T0 = T1; T1 = Tk
    end
    return val
end

function tln_error(f, N, ell)
    _, x = cheb_matrix(N)
    fv = zeros(length(x))
    ok = abs.(x) .< 1.0 - 1e-12
    y_ok = ell .* (1 .+ x[ok]) ./ (1 .- x[ok])
    fv[ok] .= f.(y_ok)
    a = dct1_coeffs(fv)
    y_fine = collect(range(0.001, 60.0, length=4001))
    x_fine = (y_fine .- ell) ./ (y_fine .+ ell)
    approx = cheb_eval(a, x_fine, N)
    return maximum(abs.(approx .- f.(y_fine)))
end

function run()
    outdir = joinpath(@__DIR__, "..", "..", "..",
                      "textbook", "figures", "ch20", "julia")
    mkpath(outdir)

    Ns = [8, 12, 16, 24, 32, 48, 64, 96]
    err_lag_e = [laguerre_error(y -> exp(-y), N) for N in Ns]
    err_lag_a = [laguerre_error(y -> 1 / (1 + y), N) for N in Ns]
    err_tln_e = [tln_error(y -> exp(-y), N, 2.0) for N in Ns]
    err_tln_a = [tln_error(y -> 1 / (1 + y), N, 5.0) for N in Ns]

    NAVY = colorant"#142D6E"; CORAL = colorant"#E74C3C"; TEAL = colorant"#16A085"
    fig = Figure(size=(1020, 340))

    ax1 = Axis(fig[1, 1], xlabel="N", ylabel="max error",
               yscale=log10, title="(a) f = e^{-y} (exponential)")
    scatterlines!(ax1, Ns, max.(err_lag_e, 1e-18); color=CORAL, label="Laguerre")
    scatterlines!(ax1, Ns, max.(err_tln_e, 1e-18); color=TEAL, label="TL_n ell=2")
    axislegend(ax1; position=:rt)

    ax2 = Axis(fig[1, 2], xlabel="N", ylabel="max error",
               yscale=log10, title="(b) f = 1/(1+y) (algebraic)")
    scatterlines!(ax2, Ns, max.(err_lag_a, 1e-18); color=CORAL, label="Laguerre")
    scatterlines!(ax2, Ns, max.(err_tln_a, 1e-18); color=TEAL, label="TL_n ell=5")
    axislegend(ax2; position=:rt)

    save(joinpath(outdir, "laguerre_vs_tln.pdf"), fig)
    save(joinpath(outdir, "laguerre_vs_tln.png"), fig)
    @printf("[20.5-julia] saved figure\n")
end

if abspath(PROGRAM_FILE) == @__FILE__
    run()
end
